### DISCUSSION TITLE
Using automapper with Prisma

### DISCUSSION THREAD

**Discussion by t0xa** [nartc/mapper#280] [Q&A]
Hello @nartc,

First of all, thank you for such incredible project. It helped me to build things much faster. Do you have a Patreon or Buy me a coffee so we can show our appreciation?

Could you please help me to understand how to correctly use Automapper with [Prisma](https://www.prisma.io/) ORM.

The main problem is that Prisma generates `type` instead of `class`. For example:

```typescript
export type Endpoint = {
  id: number
  guid: string
  description: string | null
  url: string
  status: EndpointStatus
  createdat: Date
  updatedat: Date
}
```

Is there a way of doing `mapper.createMap()` with types or some other way of creating such mapper?

Thank you in advance.

**Comment by nartc** [comment]
hi @t0xa, thanks for the kind words. I don't have Pateron or Buy me a coffee which I think I should set up (I do have Github Sponsors but people tend to stay away from Monthly fee 😛). Thanks again for your support.

As far as mapping with types/interfaces, there's [@automapper/pojos](https://github.com/nartc/mapper/blob/main/packages/pojos/README.md) but I am not too confident about it. If you could, please check out the POJOs plugin out and see if it works for you, what can be improved, what needs to be fixed etc... I'll gladly support and provide fixes.

**Comment by t0xa** [comment]
Hi @nartc, thanks a lot for suggestion. Let's close this question for now. I will open another ticket in case of issues.

May I suggest to enable Github Discussion feature so that such questions would be easier to navigate.

Regarding the sponsoring, can you do something similar to what OctoPrint does on her main page?

![image](https://user-images.githubusercontent.com/3533/111188911-38505000-85be-11eb-9b0b-4f84296cf52d.png)

**Comment by nartc** [comment]
I have just turned on Discussions. And I'll check out that OctoPrint. Thanks!

**Comment by notmedia** [comment]
I just tested Prisma types with pojos plugin and it works perfect.

**Comment by t0xa** [comment]
Thanks @notmedia. Out of curiosity, did you have any `Prisma.JsonValue` fields in your schema?

**Comment by notmedia** [comment]
Yes, in my case I have entity with JsonValue. The structure of JSON field depends on another field
So, I created class for this Json and then with automapper create an instance of that class for json field.

I need this for GraphQL, for automatic resolve type

```ts
// This type generated by Prisma
type PrismaEntity = {
  type: 'one' | 'two';
  config: JsonValue,
}

@ObjectType()
class OneConfig {
  constructor(config: { [K in keyof OneConfig]: OneConfig[K] }) {
    this.test = config.test;
  }

  @Field({
    description: 'Test value',
  })
  test: string;
}

@ObjectType()
class TwoConfig {
 ...
}

const ConfigUnion = createUnionType({
  name: 'ConfigUnion',
  types: () => [OneConfig, TwoConfig],
});


@ObjectType()
class Entity {
  @Field(...)
  type: 'one' | 'two';
  
  @Field(...)
  config: typeof ConfigUnion;
}

// Created for POJOs mapping from Prisma entity to backend entity
createMetadataMap<Entity>('Entity', {
  type: String,
});

createMetadataMap<PrismaEntity>('PrismaEntity', 'Entity', {
  type: String,
});


@Injectable()
class PrismaEntityProfile extends AutomapperProfile {
  constructor(@InjectMapper('pojos') mapper: Mapper) {
    super(mapper);
  }

  mapProfile() {
    return (mapper: Mapper) => {
      mapper.createMap<PrismaEntity, Entity>('PrismaEntity', 'Entity').forMember(
        (destination) => destination.config,
        mapFrom((source) => {
           switch(source.type) {
             case 'one': return new OneConfig(source.config as OneConfig)
             ...
           }
        })
      );
    };
  }
}
```

**Reply by nartc** [reply]
Converted to a discussion for ease of conversation. cc @t0xa  @notmedia

**Reply by linjer** [reply]
@notmedia Thanks for providing an example!

Have you figured out how to handle Prisma's `include` types? It appears to come back as something like `PrismaEntity & { relatedObj1: Obj1 }`

**Reply by notmedia** [reply]
> @notmedia Thanks for providing an example!
> 
> Have you figured out how to handle Prisma's `include` types? It appears to come back as something like `PrismaEntity & { relatedObj1: Obj1 }`

@linjer 
Hey! I moved from Prisma to another ORM. The main reason is prisma’s autogen that placing types inside node_module which we can’t modify, use decorators or provide logic. 
Prisma advantages sounds great but broke normal architecture patterns. So I cant remember how to handle it, sorry. This was a long time ago
